core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -249,29 +262,71 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
             case H1C:
             case H1:
                 // XXX(trustin): How much time is 'a little bit'?
-                ctx.channel().eventLoop().schedule(this::cleanup, 1, TimeUnit.SECONDS);
+                ch.eventLoop().schedule(() -> cleanup0(ch, shutdown, completionFuture), 1, TimeUnit.SECONDS);
                 break;
             default:
                 // HTTP/2 is unaffected by this issue because a client is expected to wait for a frame with
                 // endOfStream set.
-                cleanup();
+                cleanup0(ch, shutdown, completionFuture);
         }
     }
 
-    private void cleanup() {
+    private void cleanup0(Channel ch, boolean shutdown, @Nullable CompletableFuture<Void> completionFuture) {
+        final Throwable defaultCause = shutdown ? ShuttingDownException.get() : newClosedSessionException(ch);
         if (!unfinishedRequests.isEmpty()) {
             isCleaning = true;
-            final ClosedSessionException cause = ClosedSessionException.get();
             unfinishedRequests.forEach((req, res) -> {
                 // An HTTP2 request is cancelled by Http2RequestDecoder.onRstStreamRead()
                 final boolean cancel = !protocol.isMultiplex();
                 // Mark the request stream as closed due to disconnection.
+                Throwable cause = null;
+                if (shutdown) {
+                    cause = shutdownError(req);
+                }
+                if (cause == null) {
+                    cause = defaultCause;
+                }
                 req.abortResponse(cause, cancel);
             });
+
+            if (completionFuture != null) {
+                final CompletableFuture<?>[] futures =
+                        unfinishedRequests.keySet().stream()
+                                          .map(DecodedHttpRequest::whenResponseSent)
+                                          .toArray(CompletableFuture[]::new);
+                CompletableFuture.allOf(futures).handle((unused0, unused1) -> {
+                    completionFuture.complete(null);
```

Comment:
Note: checked that the order of completing the future vs. close the encoder doesn't really matter as 1) pipeline requests are aborted above 2) the timing of closing the `keepAliveHandler` doesn't really matter